pkg/didcomm/protocol/introduce/service.go

```diff
@@ -490,113 +482,195 @@ func isNoOp(s state) bool {
-	if msg.Invitation != nil {
-		return nil
-	}
+	var (
+		current   = md.state
+		actions   []stateAction
+		stateName string
+	)
 
-	// increases counter to determine from who we got an invitation
-	msg.IntroduceeIndex++
+	for !isNoOp(current) {
+		stateName = current.Name()
 
-	var resp = Response{}
+		next, action, err := s.execute(current, md)
+		if err != nil {
+			return fmt.Errorf("execute: %w", err)
+		}
 
-	if err := msg.Msg.Decode(&resp); err != nil {
-		return err
+		actions = append(actions, action)
+
+		if !isNoOp(next) && !current.CanTransitionTo(next) {
+			return fmt.Errorf("invalid state transition: %s --> %s", current.Name(), next.Name())
+		}
+
+		current = next
 	}
 
-	// sets an invitation to which will be forwarded later
-	msg.Invitation = resp.Invitation
+	if err := s.saveStateName(md.piID, stateName); err != nil {
+		return fmt.Errorf("failed to persist state %s: %w", stateName, err)
+	}
+
+	for _, action := range actions {
+		if action == nil {
+			continue
+		}
+
+		if err := action(); err != nil {
+			return err
+		}
+	}
 
 	return nil
 }
 
-func (s *Service) handle(msg *metaData) error {
-	current := stateFromName(msg.StateName)
+func contextInvitation(msg service.DIDCommMsg) (*didexchange.Invitation, error) {
+	var inv *didexchange.Invitation
 
-	for !isNoOp(current) {
-		next, err := s.execute(current, msg)
-		if err != nil {
-			return fmt.Errorf("execute: %w", err)
+	switch v := msg.Metadata()[metaInvitation].(type) {
+	case service.DIDCommMsgMap:
+		if err := v.Decode(&inv); err != nil {
+			return nil, err
+		}
+	case map[string]interface{}:
+		if err := service.DIDCommMsgMap(v).Decode(&inv); err != nil {
+			return nil, err
 		}
+	}
 
-		if !isNoOp(next) && !current.CanTransitionTo(next) {
-			return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
+	return inv, nil
+}
+
+type participant struct {
+	Invitation *didexchange.Invitation
+	Approve    bool
+	MessageID  string
+	MyDID      string
+	TheirDID   string
+	ThreadID   string
+}
+
+func (s *Service) saveResponse(md *metaData) error {
+	// ignore if message is not response
+	if md.Msg.Type() != ResponseMsgType {
+		return nil
+	}
+
+	// checks whether response was already handled
+	for _, p := range md.participants {
+		if p.MessageID == md.Msg.ID() {
+			return nil
 		}
+	}
 
-		current = next
+	r := Response{}
+	if err := md.Msg.Decode(&r); err != nil {
+		return err
 	}
 
-	return nil
+	thID, err := md.Msg.ThreadID()
+	if err != nil {
+		return fmt.Errorf("threadID: %w", err)
+	}
+
+	// TODO: code below might cause an issue! [ issues-1284 ]
+	//  Since the protocol support handling messages simultaneously,
+	//  there is a possibility when two requests might be executed at the same time.
+	//  It might cause a data race. As a result, some of the participants might be lost.
+	//  Possible solutions:
+	//  1. Get and save participants must be done under the lock or in the same transaction.
```

Comment:
we cannot depend on locks. we can be running in a cluster against a distributed database.